### PR TITLE
SNOW-294151 GET Implementation

### PR DESCRIPTION
**Pull request by sfc-gh-jbahk**
### Description
GET implementation for AWS, Azure and GCP

### Checklist
- [x] Code compiles correctly
- [x] Run ``make fmt`` to fix inconsistent formats
- [x] Run ``make lint`` to get lint errors and fix all of them
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

File: put_get_with_azure_test.go [hunk 1]
```diff
@@ -76,5 +79,58 @@ func testPutWithAzure(t *testing.T, isStream bool) {
 		dbt.mustExec("copy into " + tableName)
 		dbt.mustExec("rm @%" + tableName)
 		dbt.mustQueryAssertCount("ls @%"+tableName, 0)
+
+		dbt.mustExec(fmt.Sprintf(`copy into @%%%v from %v file_format=(type=csv
```
Comment (sfc-gh-cshi) [inline put_get_with_azure_test.go:83]: this piece of test code seems identical as the test added for AWS

File: put_get_with_gcp_test.go [hunk 2]
```diff
@@ -52,5 +54,58 @@ func TestPutWithGCP(t *testing.T) {
 		dbt.mustExec("copy into " + tableName)
 		dbt.mustExec("rm @%" + tableName)
 		dbt.mustQueryAssertCount("ls @%"+tableName, 0)
+
+		dbt.mustExec(fmt.Sprintf(`copy into @%%%v from %v file_format=(type=csv
```
Comment (sfc-gh-cshi) [inline put_get_with_gcp_test.go:58]: this code seems identical to other tests

File: s3_util.go [hunk 3]
```diff
@@ -188,9 +188,50 @@ func (util *snowflakeS3Util) uploadFile(
...
+	s3loc := util.extractBucketNameAndPath(meta.stageInfo.Location)
+	s3path := s3loc.s3Path + strings.TrimLeft(meta.dstFileName, "/")
+	client, ok := meta.client.(*s3.Client)
+	if !ok {
+		return &SnowflakeError{
+			Message: "failed to cast to s3 client",
+		}
+	}
+
+	f, err := os.OpenFile(fullDstFileName, os.O_CREATE|os.O_WRONLY, os.ModePerm)
```
Comment (sfc-gh-cshi) [inline s3_util.go:204]: when to close f?

File: azure_util.go [hunk 4]
```diff
@@ -165,8 +165,28 @@ func (util *snowflakeAzureUtil) uploadFile(
...
+	azureLoc := util.extractContainerNameAndPath(meta.stageInfo.Location)
+	path := azureLoc.path + strings.TrimLeft(meta.dstFileName, "/")
+	azContainerURL, ok := meta.client.(*azblob.ContainerURL)
+	if !ok {
+		return &SnowflakeError{
+			Message: "failed to cast to azure client",
+		}
+	}
+
+	f, _ := os.OpenFile(fullDstFileName, os.O_CREATE|os.O_WRONLY, os.ModePerm)
```
Comment (sfc-gh-cshi) [inline azure_util.go:181]: when do we close f?

File: gcs_util.go [hunk 5]
```diff
@@ -214,8 +210,82 @@ func (util *snowflakeGcsUtil) uploadFile(dataFile string, meta *fileMetadata, en
...
+			meta.lastError = fmt.Errorf(resp.Status)
+			meta.resStatus = renewPresignedURL
+		} else if accessToken != "" && util.isTokenExpired(resp) {
+			meta.lastError = fmt.Errorf(resp.Status)
+			meta.resStatus = renewToken
+		}
+		return meta.lastError
+	}
+
+	f, err := os.OpenFile(fullDstFileName, os.O_CREATE|os.O_WRONLY, os.ModePerm)
```
Comment (sfc-gh-cshi) [inline gcs_util.go:253]: when to close f?

File: storage_client.go [hunk 6]
```diff
@@ -157,7 +161,76 @@ func (rsu *remoteStorageUtil) uploadOneFileWithRetry(meta *fileMetadata) error {
...
+	var lastErr error
+	maxRetry := defaultMaxRetry
+	for retry := 0; retry < maxRetry; retry++ {
+		utilClass.nativeDownloadFile(meta, fullDstFileName, maxConcurrency)
+		if meta.resStatus == downloaded {
+			if meta.encryptionMaterial != nil {
+				if meta.presignedURL != nil {
+					header = utilClass.getFileHeader(meta, meta.srcFileName)
+				}
+				tmpDstFileName, err := decryptFile(header.encryptionMetadata,
```
[the remark this answers is not in the document]
Comment (sfc-gh-jbahk) [reply storage_client.go:197]: The temp file is a byproduct of the decryption when it reads from the encrypted file. I can change the decryption method to always write to a specific file name but I wanted to keep it parallel with the encryption mechanism. WDYT?

**Pull request merged**